### DISCUSSION TITLE
Barometer problem

### DISCUSSION THREAD

**Discussion by CasperLiZ** [bitcraze/discussions#914] [Q&A]
I have a question. Due to environmental factors, I use the air pressure sensor to fix the height of the CF2 aircraft and the optical flow module to fix the point. When using the hover mode, the position ring uses the laser and the speed ring uses the data obtained by merging the accelerometer and the barometer. However, Every time you start flying, it will suddenly fly very fast, causing the aircraft to crash. The setpoint Z is 0.3M. Is there anything that can be improved?

**Comment by krichardsson** [comment]
Do you mean that you are using the Flow deck and have disabled the distance sensor (the laser) and use the barometer instead for Z-estimation?

I don't have a good answer for you, but I'll do some reasoning around the topic.
The barometer is not used that much and it is not unlikely that the functionality could be improved :-)

I assume you have enabled the [`KALMAN_USE_BARO_UPDATE` define](https://github.com/bitcraze/crazyflie-firmware/blob/29f2cd6b7ee12fb85c8b281fbaafcaf31406ec29/src/modules/src/estimator/estimator_kalman.c#L100C5-L100C5)?

It looks like the barometer value is [sampled at takeoff](https://github.com/bitcraze/crazyflie-firmware/blob/29f2cd6b7ee12fb85c8b281fbaafcaf31406ec29/src/modules/src/kalman_core/kalman_core.c#L328-L331) to be used as a reference for readings later on.

```C
if (!quadIsFlying || this->baroReferenceHeight < 1) {
    this->baroReferenceHeight = baroAsl;
}
```
I'm not exactly sure what the `this->baroReferenceHeight < 1` part is doing though, i think it looks a bit fishy. I would assume that `baroAsl` could be negative in some cases (depending on the air pressure) and that would continuously update the `baroReferenceHeight`, also when flying, which feels weird to me. You could maybe try to remove that part?
 
> The setpoint Z is 0.3M. 

Have you tried to log the estimated Z (`stateEstimate.z`) to see what kind of values it has? Maybe that gives a clue to what is happening?

**Reply by puliceworx** [reply]
Apologies for joining but I tried to enable the USE_BARO_UPDATE with disastrous results.  

I also wondered what this was about this->baroReferenceHeight < 1.  

I am going to try again logging the estimateZ and messing with that reference.  
I'm just concerned with bars noise, I've measured it and it's very noisy (as expected). 
I'm using the BMP390 not the BMP388 although according to Bosch code they have the unified driver for the 3xx family.
